addons/stock_picking_batch/models/stock_move_line.py

```diff
@@ -89,3 +91,258 @@ def _add_to_wave(self, wave=False):
+        batchable_line_ids = OrderedSet()
+        for line in self:
+            if not line._is_auto_waveable():
+                continue
+            if not line.picking_type_id.wave_group_by_location:
+                batchable_line_ids.add(line.id)
+                continue
+            # We want to find the most descendant location in the wave locations list that is a parent of the line location.
+            # Since the wave locations are ordered by complete_name (from the most descendant to the most ancestor), we can iterate in reverse order.
+            for location in reversed(line.picking_type_id.wave_location_ids):
+                if line.location_id._child_of(location):
+                    lines_nearest_parent_locations[line] = location
+                    batchable_line_ids.add(line.id)
+                    break
+        batchable_lines = self.env['stock.move.line'].browse(batchable_line_ids)
+
+        remaining_line_ids = batchable_lines._auto_wave_lines_into_existing_waves(nearest_parent_locations=lines_nearest_parent_locations)
+        remaining_lines = self.env['stock.move.line'].browse(remaining_line_ids)
+        if remaining_lines:
+            remaining_lines._auto_wave_lines_into_new_waves(nearest_parent_locations=lines_nearest_parent_locations)
+
+    def _auto_wave_lines_into_existing_waves(self, nearest_parent_locations=False):
+        """ Try to add move lines to existing waves if possible, return move lines of which no appropriate waves were found to link to
+         :param nearest_parent_locations (defaultdict): the key is the move line and the value is the nearest parent location in the wave locations list"""
+        picking_types = self.mapped('picking_type_id')
+        remaining_lines = OrderedSet()
+        for picking_type in picking_types:
+            lines = self.filtered(lambda l: l.picking_type_id == picking_type)
+            if lines:
+                domain = [
+                    ('picking_type_id', '=', picking_type.id),
+                    ('company_id', 'in', lines.mapped('company_id').ids),
+                    ('is_wave', '=', True)
+                ]
+                if picking_type.batch_auto_confirm:
+                    domain = expression.AND([domain, [('state', 'not in', ['done', 'cancel'])]])
+                else:
+                    domain = expression.AND([domain, [('state', '=', 'draft')]])
+                if picking_type.batch_group_by_partner:
+                    domain = expression.AND([domain, [('picking_ids.partner_id', 'in', lines.move_id.partner_id.ids)]])
+                if picking_type.batch_group_by_destination:
+                    domain = expression.AND([domain, [('picking_ids.partner_id.country_id', 'in', lines.move_id.partner_id.country_id.ids)]])
+                if picking_type.batch_group_by_src_loc:
+                    domain = expression.AND([domain, [('picking_ids.location_id', 'in', lines.location_id.ids)]])
```

Comment:
Couldn't it be a good idea to add location_id and location_dest_id on batch for this kind of domain and the feature in general?

Reply:
What if the batch contained move lines with multiple source (or destination) locations?